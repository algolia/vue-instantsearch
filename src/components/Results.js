import algoliaComponent from '../component';

export default {
  mixins: [algoliaComponent],
  props: {
    stack: {
      type: Boolean,
      default: false,
    },
    resultsPerPage: {
      type: Number,
    },
    tagName: {
      type: String,
      default: 'div',
    },
  },
  data() {
    return {
      blockClassName: 'ais-results',
    };
  },
  render(h) {
    if (!this.show) {
      return undefined;
    }

    const children = [];
    if (this.$slots.header) {
      children.push(this.$slots.header);
    }

    this.results.forEach(result => {
      if (this.$scopedSlots.default) {
        children.push(this.$scopedSlots.default({ result }));
      } else {
        children.push(h('div', `'objectID': ${result.objectID}`));
      }
    });

    if (this.$slots.footer) {
      children.push(this.$slots.footer);
    }

    return h(this.tagName, { class: this.bem() }, children);
  },
  mounted() {
    this.updateResultsPerPage();
  },
  watch: {
    resultsPerPage() {
      this.updateResultsPerPage();
    },
  },
  methods: {
    updateResultsPerPage() {
      if (typeof this.resultsPerPage === 'number' && this.resultsPerPage > 0) {
        this.searchStore.resultsPerPage = this.resultsPerPage;
      }
    },
  },
  computed: {
    results() {
      if (this.stack === false) {
        return this.searchStore.results;
      }

      if (typeof this.stackedResults === 'undefined') {
        this.stackedResults = [];
      }

      if (this.searchStore.page === 1) {
        this.stackedResults = [];
      }

      if (
        this.stackedResults.length === 0 ||
        this.searchStore.results.length === 0
      ) {
        this.stackedResults.push(...this.searchStore.results);
      } else {
        const lastStacked = this.stackedResults[this.stackedResults.length - 1];
        const lastResult = this.searchStore.results[
          this.searchStore.results.length - 1
        ];

        if (lastStacked.objectID !== lastResult.objectID) {
          this.stackedResults.push(...this.searchStore.results);
        }
      }

      return this.stackedResults;
    },
    show() {
      return this.results.length > 0;
    },
  },
};
