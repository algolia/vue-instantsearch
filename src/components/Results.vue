<template>
  <div :class="bem()" v-if="show">

    <slot name="header"></slot>

    <slot v-for="result in results" :result="result">
      Result 'objectID': {{ result.objectID }}
    </slot>

    <slot name="footer"></slot>

  </div>
</template>

<script>import algoliaComponent from '../component';

export default {
  mixins: [algoliaComponent],
  props: {
    stack: {
      type: Boolean,
      default: false,
    },
    resultsPerPage: {
      type: Number,
    },
  },
  data() {
    return {
      blockClassName: 'ais-results',
    };
  },
  created() {
    this.updateResultsPerPage();
  },
  watch: {
    resultsPerPage() {
      this.updateResultsPerPage();
    },
  },
  methods: {
    updateResultsPerPage() {
      if (typeof this.resultsPerPage === 'number' && this.resultsPerPage > 0) {
        this.searchStore.resultsPerPage = this.resultsPerPage;
      }
    },
  },
  computed: {
    results() {
      if (this.stack === false) {
        return this.searchStore.results;
      }

      if (typeof this.stackedResults === 'undefined') {
        this.stackedResults = [];
      }

      if (this.searchStore.page === 1) {
        this.stackedResults = [];
      }

      if (
        this.stackedResults.length === 0 ||
        this.searchStore.results.length === 0
      ) {
        this.stackedResults.push(...this.searchStore.results);
      } else {
        const lastStacked = this.stackedResults[this.stackedResults.length - 1];
        const lastResult = this.searchStore.results[
          this.searchStore.results.length - 1
        ];

        if (lastStacked.objectID !== lastResult.objectID) {
          this.stackedResults.push(...this.searchStore.results);
        }
      }

      return this.stackedResults;
    },
    show() {
      return this.results.length > 0;
    },
  },
};
</script>
