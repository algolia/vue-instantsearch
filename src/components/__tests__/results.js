import Vue from 'vue';
import Results from '../Results';

test('renders proper HTML', () => {
  const searchStore = {
    results: [
      { objectID: 1 },
      { objectID: 2 },
      { objectID: 3 },
      { objectID: 4 },
    ],
  };
  const Component = Vue.extend(Results);
  const vm = new Component({
    propsData: {
      searchStore,
    },
  });
  vm.$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();
});

test('should be hidden if not results are yielded', () => {
  const searchStore = {
    results: [],
  };
  const Component = Vue.extend(Results);
  const vm = new Component({
    propsData: {
      searchStore,
    },
  });
  vm.$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();
});

test('should allow to override result rendering', () => {
  const searchStore = {
    results: [{ objectID: 1 }, { objectID: 2 }],
  };

  const Component = Vue.extend({
    render(h) {
      return h('ais-results', {
        props: { searchStore },
        scopedSlots: {
          default: ({ result }) => h('h2', result.objectID),
        },
      });
    },
    components: {
      'ais-results': Results,
    },
  });

  const vm = new Component();

  vm.$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();
});

test('can stack results', () => {
  const searchStore = {
    page: 1,
    results: [{ objectID: 1 }, { objectID: 2 }],
  };
  const Component = Vue.extend(Results);
  const vm = new Component({
    propsData: {
      searchStore,
      stack: true,
    },
  });
  vm.$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();

  vm.searchStore.page = 2;
  vm.searchStore.results = [{ objectID: 3 }, { objectID: 4 }];

  vm.$nextTick(() => {
    expect(vm.$el.outerHTML).toMatchSnapshot();
  });
});

test('should not stack results by default', () => {
  const searchStore = {
    page: 1,
    results: [{ objectID: 1 }, { objectID: 2 }],
  };
  const Component = Vue.extend(Results);
  const vm = new Component({
    propsData: {
      searchStore,
    },
  });
  vm.$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();

  vm.searchStore.page = 2;
  vm.searchStore.results = [{ objectID: 3 }, { objectID: 4 }];

  vm.$nextTick(() => {
    expect(vm.$el.outerHTML).toMatchSnapshot();
  });
});

test('should allow to customize root wrapper tag name', () => {
  const searchStore = {
    page: 1,
    results: [{ objectID: 1 }, { objectID: 2 }],
  };
  const Component = Vue.extend(Results);
  const vm = new Component({
    propsData: {
      searchStore,
      tagName: 'thead',
    },
  });
  vm.$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();
});
