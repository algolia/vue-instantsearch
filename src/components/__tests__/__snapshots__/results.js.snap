// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can stack results 1`] = `

<div class="ais-results">
  <div>
    'objectID': 1
  </div>
  <div>
    'objectID': 2
  </div>
</div>

`;

exports[`can stack results 2`] = `

<div class="ais-results">
  <div>
    'objectID': 1
  </div>
  <div>
    'objectID': 2
  </div>
  <div>
    'objectID': 3
  </div>
  <div>
    'objectID': 4
  </div>
</div>

`;

exports[`renders proper HTML 1`] = `

<div class="ais-results">
  <div>
    'objectID': 1
  </div>
  <div>
    'objectID': 2
  </div>
  <div>
    'objectID': 3
  </div>
  <div>
    'objectID': 4
  </div>
</div>

`;

exports[`should allow to customize root wrapper tag name 1`] = `

<thead class="ais-results">
  <div>
    'objectID': 1
  </div>
  <div>
    'objectID': 2
  </div>
</thead>

`;

exports[`should allow to override result rendering 1`] = `

<div class="ais-results">
  <h2>
    1
  </h2>
  <h2>
    2
  </h2>
</div>

`;

exports[`should be hidden if not results are yielded 1`] = `undefined`;

exports[`should not stack results by default 1`] = `

<div class="ais-results">
  <div>
    'objectID': 1
  </div>
  <div>
    'objectID': 2
  </div>
</div>

`;

exports[`should not stack results by default 2`] = `

<div class="ais-results">
  <div>
    'objectID': 3
  </div>
  <div>
    'objectID': 4
  </div>
</div>

`;
